File: game/src/main/java/me/ryleykimmel/brandywine/game/message/listener/LoginMessageListener.java
package me.ryleykimmel.brandywine.game.message.listener;

import io.netty.channel.ChannelFutureListener;
import me.ryleykimmel.brandywine.game.auth.AuthenticationRequest;
import me.ryleykimmel.brandywine.game.auth.AuthenticationService;
import me.ryleykimmel.brandywine.game.message.LoginMessage;
import me.ryleykimmel.brandywine.game.message.LoginResponseMessage;
import me.ryleykimmel.brandywine.game.model.World;
import me.ryleykimmel.brandywine.game.model.player.PlayerCredentials;
import me.ryleykimmel.brandywine.network.ResponseCode;
import me.ryleykimmel.brandywine.network.Session;
import me.ryleykimmel.brandywine.network.message.MessageListener;

/**
 * Listener for the {@link LoginMessage}.
 */
public final class LoginMessageListener implements MessageListener<LoginMessage> {

  /**
   * The expected client version received.
   */
  private static final int CLIENT_VERSION = 317;

  /**
   * The expected value of a dummy byte sent at the start of the login protocol.
   */
  private static final int EXPECTED_DUMMY = 255;

  /**
   * The expected opcode which indicates successful our secure RSA buffer is using the correct key pair.
   */
  private static final int EXPECTED_BLOCK_OPCODE = 10;

  /**
   * The World for submitting authentication requests.
   */
  private final World world;

  /**
   * Constructs a new {@link LoginMessageListener}.
   *
   * @param world The World for submitting authentication requests.
   */
  public LoginMessageListener(World world) {
    this.world = world;
  }

  @Override
  public void handle(Session session, LoginMessage message) {
    if (message.getDummy() != EXPECTED_DUMMY) {
      closeWithResponse(session, ResponseCode.STATUS_LOGIN_SERVER_REJECTED_SESSION);
      return;
    }

    if (message.getClientVersion() != CLIENT_VERSION) {
      closeWithResponse(session, ResponseCode.STATUS_GAME_UPDATED);
      return;
    }

    if (message.getDetail() != 0 && message.getDetail() != 1) {
      closeWithResponse(session, ResponseCode.STATUS_LOGIN_SERVER_REJECTED_SESSION);
      return;
    }

    if (message.getBlockOperationCode() != EXPECTED_BLOCK_OPCODE) {
      closeWithResponse(session, ResponseCode.STATUS_LOGIN_SERVER_REJECTED_SESSION);
      return;
    }

    long serverSessionId = message.getServerSessionId();
    long clientSessionId = message.getClientSessionId();

    if (serverSessionId != session.getId()) {
      closeWithResponse(session, ResponseCode.STATUS_BAD_SESSION_ID);
      return;
    }

    int[] sessionIds = {
        (int) (clientSessionId >> 32),
        (int) clientSessionId,
        (int) (serverSessionId >> 32),
        (int) serverSessionId
    };

    world.getService(AuthenticationService.class).submit(new AuthenticationRequest(session,
        new PlayerCredentials(message.getUserId(), message.getUsername(), message.getPassword(),
            sessionIds)));
  }

  /**
   * Closes the specified Session and notifies the client with the specified ResponseCode.
   *
   * @param session The Session to close.
   * @param response The ResponseCode sent to the client.
   */
  private void closeWithResponse(Session session, ResponseCode response) {
    session.writeAndFlush(new LoginResponseMessage(response))
        .addListener(ChannelFutureListener.CLOSE);
  }

}

File: game/src/main/kotlin/plugin/message/LoginMessageListener.kt
package plugin.message

import io.netty.channel.ChannelFutureListener
import me.ryleykimmel.brandywine.game.auth.AuthenticationRequest
import me.ryleykimmel.brandywine.game.auth.AuthenticationService
import me.ryleykimmel.brandywine.game.model.player.PlayerCredentials
import me.ryleykimmel.brandywine.game.msg.LoginMessage
import me.ryleykimmel.brandywine.game.msg.LoginResponseMessage
import me.ryleykimmel.brandywine.network.ResponseCode
import me.ryleykimmel.brandywine.network.Session
import plugin.getService
import plugin.server

class LoginMessageListener : MessageListener<Session, LoginMessage> {

    val CLIENT_VERSION = 317
    val EXPECTED_DUMMY = 255
    val EXPECTED_BLOCK_OPCODE = 10

    override fun handle(source: Session, message: LoginMessage) {
        with(message) {
            if (dummy != EXPECTED_DUMMY) {
                source.closeWithResponse(ResponseCode.STATUS_LOGIN_SERVER_REJECTED_SESSION)
                return
            }

            if (clientVersion != CLIENT_VERSION) {
                source.closeWithResponse(ResponseCode.STATUS_GAME_UPDATED)
                return
            }

            if (detail != 0 && detail != 1) {
                source.closeWithResponse(ResponseCode.STATUS_LOGIN_SERVER_REJECTED_SESSION)
                return
            }

            if (blockOperationCode != EXPECTED_BLOCK_OPCODE) {
                source.closeWithResponse(ResponseCode.STATUS_LOGIN_SERVER_REJECTED_SESSION)
                return
            }

            if (sessionKeys.size != 4) { // TODO: Rid of magic
                source.closeWithResponse(ResponseCode.STATUS_BAD_SESSION_ID)
                return
            }

            val sessionKey = sessionKeys[2].toLong() shl 32 or (sessionKeys[3].toLong() and 0xFFFFFFFFL)
            if (sessionKey !== sessionKey) {
                source.closeWithResponse(ResponseCode.STATUS_BAD_SESSION_ID)
                return
            }

            server.getService(AuthenticationService::class).submit(AuthenticationRequest(source, PlayerCredentials(userId, username, password, sessionKeys)))
        }
    }

    fun Session.closeWithResponse(response: ResponseCode) = this.writeAndFlush(LoginResponseMessage(response)).addListener(ChannelFutureListener.CLOSE)!!

}
